src/veh_interact.cpp

```diff
@@ -934,79 +928,180 @@ void veh_interact::do_mend()
+    auto sel = []( const vehicle_part &pt ) { return pt.is_tank() || pt.is_reactor(); };
+
+    auto act = [&]( const vehicle_part &pt ) {
+        auto validate = [&]( const item &obj ) {
+            if( pt.is_tank() ) { 
+                // cannot refill using active liquids (those that rot) due to #18570
+                if( obj.is_watertight_container() && !obj.contents.empty() && !obj.contents.front().active ) {
+                    return pt.can_reload( obj.contents.front().typeId() );
+                }
+            } else if( pt.is_reactor() ) {
+                return pt.can_reload( obj.typeId() );
+            }
+            return false;
+        };
+
+        target = g->inv_map_splice( validate, string_format( _( "Refill %s" ), pt.name().c_str() ), 1 );
+        if( !target ) {
+            return;
+        }
+
+        sel_vehicle_part = &pt;
+        sel_vpart_info = &pt.info();
+        sel_cmd = 'f';
         return;
-    default:
-        break; // no reason, all is well
+    };
+
+    overview( sel, act );
+    display_mode(' ');
+}
+
+void veh_interact::overview( std::function<bool(const vehicle_part &pt)> enable,
+                             std::function<void(const vehicle_part &pt)> action )
+{
+    struct part_option {
+        std::string key;
+        vehicle_part *part;
+
+        /** Can @param action be run for this entry? */
+        bool enabled;
+
+        /** Writes any extra details for this entry */
+        std::function<void( const vehicle_part &pt, WINDOW *w, int y )> details;
+    };
+
+    std::vector<part_option> opts;
+
+    std::map<std::string, std::function<void(WINDOW *, int)>> headers;
+
+    headers["TANK"] = []( WINDOW *w, int y ) {
+        trim_and_print( w, y, 1, getmaxx( w ) - 2, c_ltgray, _( "Tanks" ) );
+        right_print   ( w, y, 1, c_ltgray, _( "Contents     Qty" ) );
+    };
+    headers["BATTERY"] = []( WINDOW *w, int y ) {
+        trim_and_print( w, y, 1, getmaxx( w ) - 2, c_ltgray, _( "Batteries" ) );
+        right_print   ( w, y, 1, c_ltgray, _( "Capacity  Status" ) );
+    };
+    headers["REACTOR"] = []( WINDOW *w, int y ) {
+        trim_and_print( w, y, 1, getmaxx( w ) - 2, c_ltgray, _( "Reactors" ) );
+        right_print   ( w, y, 1, c_ltgray, _( "Contents     Qty" ) );
+    };
+    headers["TURRET"] = []( WINDOW *w, int y ) {
+        trim_and_print( w, y, 1, getmaxx( w ) - 2, c_ltgray, _( "Turrets" ) );
+        right_print   ( w, y, 1, c_ltgray, _( "Ammo     Qty" ) );
+    };
+
+    for( auto &pt : veh->parts ) {
+        if( pt.is_broken() ) {
+            continue;
+        }
+
+        if( pt.is_tank() && !pt.is_broken() ) {
+            // if tank contains something then display the contents in milliliters
+            auto details = []( const vehicle_part &pt, WINDOW *w, int y ) {
+                if( pt.ammo_current() != "null" ) {
+                    auto stack = units::legacy_volume_factor / item::find_type( pt.ammo_current() )->stack_size;
```

Comment:
This may segfault due to `stack_size` being zero. Though It's reassigned for ammo, the possibility remains.